feat: track token symbols for local faucets in CLI

tomyrd opened the pull request:
This PR is a first iteration of token symbol tracking for the CLI mentioned in #258 

## How it's done
The idea is that (faucet ID, token symbol) mappings are stored in a toml file with the structure:
```
0xa7616e9e939ba56a = "POL"
0xa031cc137adecd54 = "ETH"
0xa5d59f7ea872aeb8 = "BTC"
```

The path for the token symbol mapping file is stored in the config and can be changed by modifying that file. The file is automatically updated each time a new faucet is created locally. One of the benefits of having the mappings in a text file is that it is easy to import existing known faucets by just modifying the file.

## How it affects the user
The symbols may now be used in place of the faucet ID for commands that needed assets as inputs. For example:
```
miden send --sender 0x8b936d58cbe1a239 --target 0x8ed51db5637a4c2c --asset 10::POL --note-type private
```

If more than one match is found for the same token symbol the command should fail. The user may use the old method of specifying the faucet ID to avoid ambiguity. The token symbol is also shown in places of interest, like when prining the account details or it's vault.

## What's missing
Some features for future PRs:
- Support for public faucet accounts.
- Support for private faucet accounts (I don't know if this is currently possible).

bobbinth commented:
Not a review yet - but I like the overall approach. A couple of preliminary comments.

> The idea is that (faucet ID, token symbol) mappings are stored in a toml file with the structure:
> 
> ```
> 0xa7616e9e939ba56a = "POL"
> 0xa031cc137adecd54 = "ETH"
> 0xa5d59f7ea872aeb8 = "BTC"
> ```

I would probably make the structure of the TOML file a bit more explicit. Maybe something like:
```toml
POL = { faucet_id = "0xa7616e9e939ba56a" }
ETH = { faucet_id = "0xa031cc137adecd54" }
BTC = { faucet_id = "0xa5d59f7ea872aeb8" }
```

Also, I wonder if this should be just a section in the config file - e.g.,
```toml
default_account_id = "0x9822dacfb299959d"

[rpc]
timeout_ms = 10000

[rpc.endpoint]
protocol = "https"
host = "18.203.155.106"
port = 57291

[store]
database_filepath = "store.sqlite3"

[token-symbols]
POL = { faucet_id = "0xa7616e9e939ba56a" }
ETH = { faucet_id = "0xa031cc137adecd54" }
BTC = { faucet_id = "0xa5d59f7ea872aeb8" }
```
Though, there are some pros and cons to this. The biggest con is that asset symbol map may need to be distributed separately from the main config file.

Lastly, and probably not for this PR, I wonder what could be the mechanism for updating/distributing this file. I'm actually curious how this work in Ethereum and other ecosystems.

igamigo commented on bin/miden-cli/src/commands/account.rs:
```diff
@@ -227,14 +236,21 @@ pub fn show_account<N: NodeRpcClient, R: FeltRng, S: Store, A: TransactionAuthen
+            let token_symbol_mappings = get_token_mappings()?;
+            let token_symbol = token_symbol_mappings
+                .get_token_symbol(account_id)
+                .map_err(|err| format!("Failed to get token symbol: {}", err))
```
This `map_err` can probably be removed if we do `unwrap_or()` afterward

igamigo commented on bin/miden-cli/src/commands/new_transactions.rs:
```diff
@@ -95,7 +95,7 @@ pub struct SendCmd {
     target_account_id: String,
 
-    /// Asset in the format `<AMOUNT>::<FAUCET_ID_HEX>`
+    /// Asset in the format `<AMOUNT>::<FAUCET_ID_HEX>` or `<AMOUNT>::<TOKEN_SYMBOL>`
```
Should we make it clear that these `TOKEN_SYMBOL`s are expected to be found in the toml file?

igamigo reviewed (commented):
Leaving some comments for now. Regarding leaving this data in a separate file, I think it's fine (and I slightly prefer it) but I don't mind if we want to move it to the same `miden-client.toml`.

tomyrd commented:
If we change the file structure to use the token symbol as the key:
```
POL = { faucet_id = "0xa7616e9e939ba56a" }
ETH = { faucet_id = "0xa031cc137adecd54" }
BTC = { faucet_id = "0xa5d59f7ea872aeb8" }
```

How do you think we should handle symbol collisions? 
One posibility is to overwrite when a match happens but that would lead to lost information. Another option would be to allow for collisions and store the mappings like:
```
POL = { faucet_ids = ["0xa7616e9e939ba56a"] }
ETH = { faucet_ids = ["0xa031cc137adecd54", "0xa321cc137adec123"] }
BTC = { faucet_ids = ["0xa5d59f7ea872aeb8"] }
```
In this case if the user tries to send `10::ETH` the command would fail, they would need to specify the faucet ID

bobbinth commented:
> How do you think we should handle symbol collisions?

I think it should be an error if we have two faucet IDs mapping to the same symbol. Similarly, if one faucet ID maps to the same symbol, it should be an error as well. I think these should be caught when we parse the file.

One other thing that we didn't take into account is decimal places. I think the mapping could look like so:
```toml
POL = { faucet_id = "0xa7616e9e939ba56a", decimals = 6 }
ETH = { faucet_id = "0xa031cc137adecd54", decimals = 8 }
BTC = { faucet_id = "0xa5d59f7ea872aeb8", decimals = 5 }
```

This way, we could say something like `10.5::POL` and it would be translated to the right number of units.

tomyrd commented:
Added the changes mentioned in the previous comments. I didn't end up merging the config and mapping files as it is not fully decided (I also feel like they should be separate but I can merge them if we don't want to).

> Lastly, and probably not for this PR, I wonder what could be the mechanism for updating/distributing this file. I'm actually curious how this work in Ethereum and other ecosystems.

I didn't find a lot of information about this online. In Ethereum there are third party explorers (like [etherscan](https://etherscan.io/tokens)) that constantly analyze deployed contracts to see if they match with known token contracts (like ERC-20) and they extract the token symbols from there. There are also community driven aproaches like [Tokenlistooor](https://github.com/SmolDapp/tokenLists/tree/main) that compile and publish token lists. If we want to support official lists we need a way to gather all of the mappings and publish them.

> This way, we could say something like 10.5::POL and it would be translated to the right number of units.

I added the decimals to the mappings file but currently the cli doesn't support decimal places in the asset commands, should I add it in this PR?

bobbinth commented on bin/miden-cli/src/token_symbol_mappings.rs:
```diff
@@ -0,0 +1,93 @@
+            .map(|(symbol, _)| symbol.clone()))
+    }
+
+    pub fn set_token_symbol(
```
Are we updating the file dynamically somewhere? I was thinking it would be static and we'd modify it manually when needed.

igamigo replied:
I think this function is being used to add a token mapping if a faucet is created.

tomyrd replied:
Yes, each time a new local faucet is created the mapping is added to the file. This happens [here](https://github.com/0xPolygonMiden/miden-client/blob/d5eb29fd7e665fcbc10ba24af2b35d65e4e3921b/bin/miden-cli/src/commands/new_account.rs#L71-L72).

bobbinth replied:
I think for local faucets it is OK to assume that the user would edit the file manually as well. I'm assuming these will be pretty rare

One thing we can do is print out a short sentence about this after a faucet is created in the CLI.

bobbinth commented on bin/miden-cli/src/config.rs:
```diff
@@ -7,17 +7,21 @@ use figment::{
     /// Address of the Miden node to connect to.
     pub default_account_id: Option<String>,
+    /// Path to the file containing token symbol mappings.
+    pub token_symbol_mappings_file: String,
```
nit: I'd probably name this `token_symbol_map_filepath`.

igamigo replied:
Should this be of type `Path`?

bobbinth reviewed (commented):
Looks good! Thank you! Not a full review - but I did leave a couple of comments inline (the main one about potentially not updating the mappings file dynamically).

>  I didn't end up merging the config and mapping files as it is not fully decided

Agreed. Let's keep it the way you have it.

> I didn't find a lot of information about this online. In Ethereum there are third party explorers (like [etherscan](https://etherscan.io/tokens)) that constantly analyze deployed contracts to see if they match with known token contracts (like ERC-20) and they extract the token symbols from there. There are also community driven aproaches like [Tokenlistooor](https://github.com/SmolDapp/tokenLists/tree/main) that compile and publish token lists. If we want to support official lists we need a way to gather all of the mappings and publish them.

Thank you for looking into this! I think for now we'll need to probably come up with some default mappings that ship with the client - but figuring how exactly it would work is a separate issue.

> I added the decimals to the mappings file but currently the cli doesn't support decimal places in the asset commands, should I add it in this PR?

No - let's add support for decimals in a follow-up PR (and we should create an issue for this).

igamigo commented on CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 
 ## v0.5.0 (TBD)
 
+* Track token symbols for local faucet accounts (#441).
```
nit: These are usually written in past tense (there another log line two lines down that seems to need fixing as well)

igamigo commented on bin/miden-cli/src/commands/account.rs:
```diff
@@ -145,20 +145,31 @@ pub fn show_account<N: NodeRpcClient, R: FeltRng, S: Store, A: TransactionAuthen
+                    fungible_asset.faucet_id(),
+                    token_symbol_mappings
+                        .get_token_symbol(&fungible_asset.faucet_id())?
+                        .unwrap_or("Unknown".to_string()),
```
This `get_token_symbol().unwrap_or()` could be its own function (maybe something like `TokenSymbolMapping::get_name_or_default()`?)so as to not repeat it.

igamigo commented on bin/miden-cli/src/utils.rs:
```diff
@@ -171,3 +176,9 @@ pub fn build_swap_tag(
 }
+
+/// Returns the token symbol mappings from the config file.
+pub fn get_token_mappings() -> Result<TokenSymbolMappings, String> {
```
nit: I'd name this `load_token_mappings()`

igamigo reviewed (approved):
Awesome! Left some comments but in general LGTM!

bobbinth commented on bin/miden-cli/src/config.rs:
```diff
@@ -7,17 +8,21 @@ use figment::{
 use miden_client::{config::RpcConfig, store::sqlite_store::config::SqliteStoreConfig};
 use serde::{Deserialize, Serialize};
 
+const TOKEN_SYMBOL_MAPPINGS_FILE_NAME: &str = "token_symbol_mappings.toml";
```
nit: I would change this also to just `token_symbol_map.toml`

bobbinth commented on bin/miden-cli/src/token_symbol_map.rs:
```diff
@@ -0,0 +1,62 @@
+}
+
+impl TokenSymbolMap {
+    pub fn new(mappings_file: PathBuf) -> Self {
+        Self { mappings_file }
+    }
```
A slightly alternative approach could be to load the map in the constructor (i.e., the constructor basically becomes `load_mappings()` and the struct then becomes:
```rust
pub struct TokenSymbolMap(BTreeMap<String, FaucetDetails>);
```
One nice thing about this is that `get_token_symbol()` could just return `Option<String>`.

But also, if there are situations when we instantiate `TokenSymbolMap` but never use it, this would create extra work.

bobbinth commented on bin/miden-cli/src/token_symbol_map.rs:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    fn load_mappings(&self) -> Result<BTreeMap<String, FaucetDetails>, String> {
+        let mappings = match std::fs::read_to_string(&self.mappings_file) {
+            Ok(content) => match toml::from_str(&content) {
+                Ok(mappings) => mappings,
+                Err(err) => return Err(format!("Failed to parse mappings file: {}", err)),
+            },
+            Err(err) => {
+                if err.kind() != std::io::ErrorKind::NotFound {
+                    return Err(format!("Failed to read mappings file: {}", err));
+                }
+                BTreeMap::new()
+            },
+        };
```
Do we not need to validate somewhere here that the `symbol |-> faucet_id` mappings are unique?

tomyrd replied:
The uniqueness of the token symbols is checked when reading the file with `toml::from_str`. I added a check to validate that the faucet IDs are also unique.

bobbinth reviewed (commented):
Thank you! Looks good! I left a few more small comments inline.

bobbinth reviewed (approved):
Looks good! Thank you!

Let's create an issue to add decimal support to the CLI inputs.

tomyrd commented:
Done! I created #448 to track the issue.
